frame/curve-amm/src/lib.rs

```diff
@@ -1011,53 +656,139 @@ pub mod pallet {
 			None
 		}
 
-		fn transfer_liquidity_into_pool(
-			pool_account_id: &T::AccountId,
-			pool: &mut PoolInfo<T::AccountId, T::AssetId, T::Balance>,
-			source: &T::AccountId,
-			destination_asset_index: usize,
-			amount: T::Balance,
-		) -> DispatchResult {
-			T::LpToken::transfer(
-				pool.assets[destination_asset_index],
-				source,
-				pool_account_id,
-				amount,
-				true,
-			)?;
-
-			pool.total_balances[destination_asset_index] = pool.total_balances
-				[destination_asset_index]
-				.checked_add(&amount)
-				.ok_or(Error::<T>::InconsistentStorage)?;
-
-			Ok(())
-		}
-
-		fn transfer_liquidity_from_pool(
-			pool_account_id: &T::AccountId,
-			pool: &mut PoolInfo<T::AccountId, T::AssetId, T::Balance>,
-			source_asset_index: usize,
-			destination: &T::AccountId,
-			amount: T::Balance,
-		) -> DispatchResult {
-			T::LpToken::transfer(
-				pool.assets[source_asset_index],
-				pool_account_id,
-				destination,
-				amount,
-				true,
-			)?;
-
-			pool.total_balances[source_asset_index] = pool.total_balances[source_asset_index]
-				.checked_sub(&amount)
-				.ok_or(Error::<T>::InconsistentStorage)?;
-
-			Ok(())
-		}
-
 		pub fn account_id(pool_id: &PoolId) -> T::AccountId {
 			PALLET_ID.into_sub_account(pool_id)
 		}
+
+        fn limit_preserving_invariant( 
+			pool_id: PoolId,
+            index : usize,
+            min_balance : T::Balance,
+            max_balance : T::Balance,
+        ) -> Result<T::Balance, DispatchError> {
+            let value = Pools::<T>::try_mutate(pool_id, |pool| -> Result<T::Balance, DispatchError> {
+                let pool = pool.as_mut().ok_or(Error::<T>::PoolNotFound)?;
```

Comment:
i see, so caller 
1. gets balance of one asset
2. get balance of all assets
3. calculate.
4. repeat for each from one.

could we have got
1. get all balances
2. get all withdrawables/depositbles for each
3. sort from smalles balance to largest
4. withdraw max to smalles at index i = 0
5. check if can withdraw for i+1 to match i
6. if can, add it i+i to match i
7. if not, return back, cut i to be as small as i+1
8. i+i, can added to match, add, cannot <- try cut previous increases.
9. as soon as we hit something in i+n, check if has withdraw to became smaller than previous, make it smaller up to value
10. if it cannot be made as small as previous, make it as small as poossible
11. keep track of all additionas and removals
12. so we maximixed greed to take all possible and make all as equal as possible.
13. next, take a look if can withdraw without violating invariant from each
13. if each withdrawl does not violates invariant, withdraw from each as much as possible.
13. example , 4 + 2 , 7 - 2, 8 - 2 -> 6, 6, 6
14. example,  5 - 2 , 7 - 3, 8 - 5 -> 4, 4, 4
15. example,  5 + 1, 7 + 1, 10 - 1 -> 6, 7, 9
16. example,  5 - 3 , 7 , 8 - 5 -> 5, 7, 7 (incetives to add more to balance, as smalles is most imbalanced, or withdraw)
17. write property test generating random values, ensue that if you drop VERY big imblaances, it converges with smaller randoms to invarint.
18. prove that trying to add big imblances does not leads to inblanaces
19. not sure if above algo works, but it is at least more optimal than looping, may be.
20. i will with about using average or error to find most balanced solution (solution which leads to smalles possible arbitrage ability, i mean that if after balancing will trade in cyrpto, smalles possible diff amid values will be visible)

Reply:
@KaiserKarel can you check if we are not doing here too stupid or complicated. Should we rely on vault admin to balance invariant or like? Or pool should help itself?

Reply:
@vivekvpandya i pretty robust about algorithm. can meet and explain it. but first may be we could ask somebody from financial team about feacibility of such?

Reply:
I am struggling with next scneario right now

9  9 9 11 11 11 12
2  0 0  0   0  0   0      

It is evident that adding to 2 nine is right. That is why binary search and calculate X.

9  9 9 11 11 11 12
10  0 0  0   0  0   0      

In this scenario, algorithm tells, try 9 + 3. If improve -> Add 3. If not, 3/2 and check. For few numbers it can be fast even if n * n . Alternatively, can do best, with up to binary search, and do tune off chain more complex. Again, if violation of target to far from good.

So can run different variants of next, like `12  9 9 11 11 11 12` ``11  9 9 11 11 11 12`

9  9 9 11 11 11 12
10  0 7  0   0  0   0      

binary earch would be
take 7 as max. 
add 7 to 9 and to 9 -> bad.

3.5 -> add. Just like try to increase all currencies to make them equal(thismakes XI good).



And this. And see what is make to make XI = A with up to some 0.95 still making DEX leverage largest. 



![image](https://user-images.githubusercontent.com/757125/144656694-e0adc560-0d46-4185-bb32-2316492cd1b3.png)


@vivekvpandya for prove we would use random+property tests. We will run set of randomized experiments and see that algorithm ends up always to best X. 

@hussein-aitlahcen it is intresting, that if USD = 1 and DAI = 1, and ZXC = 100. Still can use ZXC if to divied it by 100. So as soon as some LP token in stable, we can create markets for it with other stable by normalizing.